### PR TITLE
Add bundle and charts to `copy packages`

### PR DESCRIPTION
**Pull request by TerryHowe**
I removed copy package bundles and helm charts from the previous PR to reduce the complexity of the last PR. The previous PR only copies the images. This PR adds the bundles and charts to the copy.
* Deprecates --include-packages and removes support from import images (won't work anyway)
* Deprecates --include-package and removes support from download images (won't work anyway)
* Meeting test coverage criteria was not easy; does not entirely make sense to me how that is calculated.

### CODE REVIEW

File: pkg/curatedpackages/reader.go [hunk 1]
```diff
@@ -28,16 +30,25 @@ type ManifestReader interface {
...
 		ManifestReader: mr,
+		AllImages:      true,
 	}
 }
 
 func (r *PackageReader) ReadImagesFromBundles(ctx context.Context, b *releasev1.Bundles) ([]releasev1.Image, error) {
-	images, err := r.ManifestReader.ReadImagesFromBundles(ctx, b)
+	var err error
+	var images []releasev1.Image
+	if r.AllImages {
```
Comment (ewollesen) [inline pkg/curatedpackages/reader.go:49]: It's unclear what the difference is between ReadImagesFromBundles and fetchImagesFromBundles.

I'm guessing the former are for curated packages, while the latter are for EKS-A core images?

If so, perhaps some renaming could make it clearer what's going on:

```
r.AllImages => r.IncludePackagesImages
r.ManifestReader.ReadImagesFromBundles => ReadPackagesImagesFromBundles
```

File: pkg/curatedpackages/reader.go [hunk 2]
```diff
@@ -53,29 +64,46 @@ func (r *PackageReader) ReadImagesFromBundles(ctx context.Context, b *releasev1.
...
+		images = append(images, eksaCharts...)
+	}
+
+	// Generate bundle repository
+	for _, chart := range eksaCharts {
+		if strings.Contains(chart.Repository(), "eks-anywhere-packages") {
+			var bundleRepository releasev1.Image
+			chart.DeepCopyInto(&bundleRepository)
+			bundleRepository.URI = strings.Replace(bundleRepository.URI, "eks-anywhere-packages", "eks-anywhere-packages-bundles", 1)
+			bundleRepository.URI = strings.Replace(bundleRepository.URI, bundleRepository.Version(), "v1-24-latest", 1)
```
Comment (ewollesen) [inline pkg/curatedpackages/reader.go:83]: Hard coding this value will likely come back to bite us.

File: pkg/curatedpackages/reader.go [hunk 3]
```diff
@@ -53,29 +64,46 @@ func (r *PackageReader) ReadImagesFromBundles(ctx context.Context, b *releasev1.
...
+			bundleRepository.ImageDigest = ""
```
Comment (ewollesen) [inline pkg/curatedpackages/reader.go:84]: Minor: this feels like a nice function to extract.

File: pkg/curatedpackages/reader.go [hunk 4]
```diff
@@ -122,13 +151,29 @@ func (r *PackageReader) fetchImagesFromBundle(ctx context.Context, versionsBundl
...
 				URI:         fmt.Sprintf("%s/%s@%s", getRegistry(ctrl.URI), version.Repository, version.Digest),
+				ImageDigest: version.Digest,
 			}
 			images = append(images, image)
 		}
 	}
 	return images, nil
 }
 
+func removeDuplicateImages(strSlice []releasev1.Image) []releasev1.Image {
```
Comment (ewollesen) [inline pkg/curatedpackages/reader.go:162]:
```suggestion
func removeDuplicateImages(images []releasev1.Image) []releasev1.Image {
```

File: pkg/curatedpackages/reader.go [hunk 5]
```diff
@@ -122,13 +151,29 @@ func (r *PackageReader) fetchImagesFromBundle(ctx context.Context, versionsBundl
...
+	allKeys := make(map[string]releasev1.Image)
+	var list []releasev1.Image
```
Comment (ewollesen) [inline pkg/curatedpackages/reader.go:164]:
```suggestion
	var uniqImages []releasev1.Image
```

File: pkg/curatedpackages/reader.go [hunk 6]
```diff
@@ -122,13 +151,29 @@ func (r *PackageReader) fetchImagesFromBundle(ctx context.Context, versionsBundl
...
+				ImageDigest: version.Digest,
 			}
 			images = append(images, image)
 		}
 	}
 	return images, nil
 }
 
+func removeDuplicateImages(strSlice []releasev1.Image) []releasev1.Image {
+	allKeys := make(map[string]releasev1.Image)
```
Comment (ewollesen) [inline pkg/curatedpackages/reader.go:163]: Nit: there's no need to copy the images into the map. You could just use:
```go
map[string]struct{}
```
Which wouldn't have to copy as many things around, saving both memory and time.
That would change line 167 to something more like: 
```go
allKeys[item.VersionedImage()] = struct{}{}
```

File: pkg/curatedpackages/reader.go (hunk 1 above)
Comment (TerryHowe) [reply pkg/curatedpackages/reader.go:50]: I agree this code is odd, but I don't want to refactor it.

File: pkg/curatedpackages/reader.go (hunk 5 above)
Comment (TerryHowe) [reply pkg/curatedpackages/reader.go:165]: Done.

File: pkg/curatedpackages/reader.go (hunk 4 above)
Comment (TerryHowe) [reply pkg/curatedpackages/reader.go]: Done.

File: pkg/curatedpackages/reader.go (hunk 6 above)
Comment (TerryHowe) [reply pkg/curatedpackages/reader.go]: Done.

File: pkg/curatedpackages/reader.go (hunk 3 above)
Comment (TerryHowe) [reply pkg/curatedpackages/reader.go:86]: Another time perhaps.

File: pkg/curatedpackages/reader.go (hunk 2 above)
Comment (TerryHowe) [reply pkg/curatedpackages/reader.go:85]: Yes.

Comment (TerryHowe) [reply pkg/curatedpackages/reader.go]: Fixed

File: cmd/eksctl-anywhere/cmd/downloadimages.go [hunk 7]
```diff
@@ -43,7 +41,8 @@ func init() {
 		log.Fatalf("Cannot mark 'output' flag as required: %s", err)
 	}
 
-	downloadImagesCmd.Flags().BoolVar(&downloadImagesRunner.includePackages, "include-packages", false, "Flag to indicate inclusion of curated packages in downloaded images")
+	downloadImagesCmd.Flags().BoolVar(&downloadImagesRunner.includePackages, "include-packages", false, "Flag no long works, use copy packages")
```
Comment (ewollesen) [inline cmd/eksctl-anywhere/cmd/downloadimages.go:44]:
```suggestion
	downloadImagesCmd.Flags().BoolVar(&downloadImagesRunner.includePackages, "include-packages", false, "this flag no longer works, use copy packages instead")
```

File: cmd/eksctl-anywhere/cmd/import_images.go [hunk 8]
```diff
@@ -48,6 +48,7 @@ func init() {
 		log.Fatalf("Cannot mark 'bundles' as required: %s", err)
 	}
 	importImagesCmd.Flags().BoolVar(&importImagesCommand.includePackages, "include-packages", false, "Flag to indicate inclusion of curated packages in imported images")
+	importImagesCmd.Flag("include-packages").Deprecated = "Use copy packages command"
```
Comment (ewollesen) [inline cmd/eksctl-anywhere/cmd/import_images.go:51]:
```suggestion
	importImagesCmd.Flag("include-packages").Deprecated = "use copy packages command"
```
Just for consistency with the other deprecation on line 45 of downloadimages.go

**Review by ewollesen — COMMENTED** [review]
/lgtm

File: pkg/curatedpackages/reader.go [hunk 9]
```diff
@@ -28,105 +30,119 @@ type ManifestReader interface {
 
 type PackageReader struct {
 	ManifestReader
+	AllImages bool
```
[the remark this answers is not in the document]
Comment (TerryHowe) [reply pkg/curatedpackages/reader.go:33]: Deleted.

File: pkg/curatedpackages/packagereader_test.go [hunk 10]
```diff
@@ -2,183 +2,166 @@ package curatedpackages_test
...
+	bundles.Spec.VersionsBundles[0].KubeVersion = "1"
 
 	images, err := tt.command.ReadImagesFromBundles(tt.ctx, bundles)
 
 	tt.Expect(err).To(BeNil())
 	tt.Expect(images).To(BeEmpty())
 }
 
-func TestPackageReaderReadImagesFromBundlesFailWhenWrongBundle(t *testing.T) {
+func TestPackageReader_ReadImagesFromBundlesBadRegistry(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (TerryHowe) [reply pkg/curatedpackages/packagereader_test.go:99]: Yeh, it just hits the continue on failures and logs a message. I do not want to change that behavior right now.

File: pkg/curatedpackages/reader.go [hunk 11]
```diff
@@ -28,105 +30,117 @@ type ManifestReader interface {
...
-	ctrl := versionsBundle.PackageController.Controller
-	bundle := &packagesv1.PackageBundle{}
-	err = yaml.Unmarshal(data, bundle)
+	bundle := packagesv1.PackageBundle{}
+	err = yaml.Unmarshal(data, &bundle)
 	if err != nil {
-		return nil, err
+		return "", nil, err
 	}
+	return artifact.VersionedImage(), &bundle, nil
```
Comment (a-cool-train) [inline pkg/curatedpackages/reader.go:97]: What is the artifact.VersionedImage() for? Wouldn't returning just the bundle and the error suffice for this method?

Comment (TerryHowe) [reply pkg/curatedpackages/reader.go:97]: The bundle does not contain the bundles location.

**Pull request merged**